File: cpp/200/Number_of_Islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int ret = 0;
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[i].size();j++){
                if(grid[i][j]=='1'){
                    solve(grid, i, j);
                    ret++;
                }
            }
        }
        return ret;
    }
    
    void solve(vector<vector<char>> &grid, int i, int j){
        grid[i][j] = '0';
        int dx[4] = {-1,0,1,0}, dy[4] = {0,-1,0,1};
        for(int k=0;k<4;++k){
            int nx = i+dx[k], ny = j+dy[k];
            if(nx>=0&&nx<grid.size()&&ny>=0&&ny<grid[nx].size()&&grid[nx][ny]=='1'){
                solve(grid, nx, ny);
            }
        }
    }
};

File: python/200/Number_of_Islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid:
            return 0
        row, col = len(grid), len(grid[0])
        dx = [1, 0, -1, 0]
        dy = [0, 1, 0, -1]
        ret = 0
        for i in range(row):
            for j in range(col):
                if int(grid[i][j]):
                    q = []
                    ret += 1
                    # BFS
                    # 将根结点压入队列
                    q.append((i, j))
                    # 开始循环
                    while len(q) > 0:
                        # 取出一个节点
                        x, y = q.pop()
                        grid[x][y] = '0'
                        # 放入周围节点
                        for k in range(4):
                            nx, ny = x + dx[k], y + dy[k]
                            if 0<=nx<row and 0<=ny<col and grid[nx][ny]=='1':
                                q.append((nx, ny))
        return ret
